protoc-gen-swagger/defs.bzl

```diff
@@ -0,0 +1,79 @@
+def _collect_includes(srcs):
+    includes = ["."]
+    for src in srcs:
+        include = src.dirname
+        if include and not include in includes:
+            includes += [include]
+
+    return includes
+
+def _run_proto_gen_swagger(direct_proto_srcs, transitive_proto_srcs, actions, protoc, protoc_gen_swagger):
+    swagger_files = []
+    for proto in direct_proto_srcs:
+      swagger_file = actions.declare_file(
+          "%s.swagger.json" % proto.basename[:-len(".proto")],
+          sibling = proto,
+      )
+
+      args = actions.args()
+      args.add("--plugin=%s" % protoc_gen_swagger.path)
+      args.add("--swagger_out=logtostderr=true:%s" % swagger_file.dirname)
+      args.add("-Iexternal/com_google_protobuf/src")
+      args.add("-Iexternal/com_github_googleapis_googleapis")
+      args.add(["-I%s" % include for include in _collect_includes(direct_proto_srcs + transitive_proto_srcs)])
+      args.add(proto.basename)
+
+      actions.run(
+          executable = protoc,
+          inputs = direct_proto_srcs + transitive_proto_srcs + [protoc_gen_swagger],
+          outputs = [swagger_file],
+          arguments = [args],
+      )
+
+      swagger_files.append(swagger_file)
+
+    return swagger_files
+
+def _proto_gen_swagger_impl(ctx):
+    direct_sources = []
+    transitive_proto_srcs = depset(ctx.files._well_known_protos)
+    for dep in ctx.attr.deps:
+        direct_sources += dep.proto.direct_sources
+        transitive_proto_srcs = depset(transitive=[transitive_proto_srcs, dep.proto.transitive_sources])
+
+    return struct(
+        files=depset(
+            _run_proto_gen_swagger(
+                direct_sources,
+                transitive_proto_srcs.to_list(),
+                ctx.actions,
+                ctx.executable._protoc,
+                ctx.executable._protoc_gen_swagger,
+            )
+        )
+    )
+
+protoc_gen_swagger = rule(
+    attrs = {
+        "deps": attr.label_list(
```

Comment:
I still think this should be called `proto`, not `deps`. You also don't need it to be a `label_list`.